Mathlib/Data/ListM/Basic.lean

```diff
@@ -103,15 +126,33 @@ unsafe def force {α} : ListM m α → m (List α)
 #align tactic.mllist.force ListM.force
 
 /-- Take the first `n` elements, as a list inside the monad. -/
-unsafe def take {α} : ListM m α → Nat → m (List α)
+unsafe def takeAsList {α} : ListM m α → Nat → m (List α)
   | nil, _ => pure []
   | _, 0 => pure []
   | cons l, n + 1 => do
     let (x, xs) ← l
     let some x ← pure x |
-      take xs (n + 1)
-    (· :: ·) x <$> take xs n
-#align tactic.mllist.take ListM.take
+      takeAsList xs (n + 1)
+    (· :: ·) x <$> takeAsList xs n
+#align tactic.mllist.take ListM.takeAsList
+
+/-- Take the first `n` elements. -/
+unsafe def take : ListM m α → Nat → ListM m α
+  | _, 0 => empty
+  | nil, _ => empty
+  | cons l, n+1 => cons do
+    let (a, l) ← l
+    let some a ← pure a | return (none, l.take (n+1))
+    return (a, l.take n)
```

Comment:
These are probably written more easily using `uncons`:
```suggestion
  | _, 0 => empty
  | xs, n+1 => cons do
    match ← uncons xs with
    | some (x, xs) => return (x, xs.take n)
    | none => return (none, empty)
```

Reply:
Oh, yes. I've rewritten all the functions that benefit from this.